lighthouse-core/audits/seo/font-size.js

```diff
@@ -0,0 +1,233 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const parseURL = require('url').parse;
+const Audit = require('../audit');
+const ViewportAudit = require('../viewport');
+const CSSStyleDeclaration = require('../../lib/web-inspector').CSSStyleDeclaration;
+const MINIMAL_PERCENTAGE_OF_LEGIBLE_TEXT = 75;
+
+/**
+ * @param {Array<{cssRule: WebInspector.CSSStyleDeclaration, fontSize: number, textLength: number, node: Node}>} fontSizeArtifact
+ * @returns {Array<{cssRule: WebInspector.CSSStyleDeclaration, fontSize: number, textLength: number, node: Node}>}
+ */
+function getUniqueFailingRules(fontSizeArtifact) {
+  const failingRules = new Map();
+
+  fontSizeArtifact.forEach(({cssRule, fontSize, textLength, node}) => {
+    const artifactId = getFontArtifactId(cssRule, node);
+    const failingRule = failingRules.get(artifactId);
+
+    if (!failingRule) {
+      failingRules.set(artifactId, {
+        node,
+        cssRule,
+        fontSize,
+        textLength,
+      });
+    } else {
+      failingRule.textLength += textLength;
+    }
+  });
+
+  return failingRules.valuesArray();
+}
+
+/**
+ * @param {Node} node
+ * @return {{type:string, snippet:string}}
+ */
+function nodeToTableNode(node) {
+  const attributesString = node.attributes.map((value, idx) =>
+    (idx % 2 === 0) ? ` ${value}` : `="${value}"`
+  ).join('');
+
+  return {
+    type: 'node',
+    snippet: `<${node.localName}${attributesString}>`,
```

Comment:
can we add a selector here so they title won't be `undefined` or do we not really have access to it at this point?

![image](https://user-images.githubusercontent.com/2301202/33564067-76c42fbe-d8ce-11e7-83ef-6dc0d52e56cb.png)

Reply:
Getting a nice looking unique selector is [quite complicated](https://github.com/dequelabs/axe-core/blob/develop/lib/core/utils/get-selector.js) and since we don't have real DOM nodes here but rather devtools protocol nodes, it'd be tricky to reuse Axe's getSelector. I've put together a [simpler alternative](https://gist.github.com/kdzwinel/2e2f35af56d6da07fbfac12069bf5cff) that would work here, but it doesn't guarantee uniques. WDTY?

BTW I realized that two of my audits that already landed: hreflang and is-crawlable, both suffer from the same issue (selector is not provided). For both of those however I don't have access to the parent nodes making it impossible to produce an useful selector. Should I just pass an empty string there?

Reply:
Hm, yeah maybe not worry about it just yet then. 

Just use the ID/class of the immediate node if we have it, add a TODO, and we'll file an issue linking off to that gist to make better selectors for these SEO audits then

Reply:
Done, PTAL: https://github.com/GoogleChrome/lighthouse/pull/3533/commits/88091d7c6aabeae6493ed5b57655b619de5d026b . Should I file the issue for that TODO?

Reply:
looks good! and yeah that'd be great :)

Reply:
https://github.com/GoogleChrome/lighthouse/issues/4015 👍